client/task_runner.go

```diff
@@ -356,25 +423,287 @@ func (r *TaskRunner) validateTask() error {
+				return
+			}
+
+			// Write the token to disk
+			if err := r.writeToken(token); err != nil {
+				e := fmt.Errorf("failed to write Vault token to disk")
+				r.logger.Printf("[ERR] client: %v for task %v on alloc %q: %v", e, r.task.Name, r.alloc.ID, err)
+				r.Kill("vault", e.Error())
+				return
+			}
+		}
+
+		// Start the renewal process
+		renewCh, err := r.vaultClient.RenewToken(token, 30)
+
+		// An error returned means the token is not being renewed
+		if err != nil {
+			r.logger.Printf("[ERR] client: failed to start renewal of Vault token for task %v on alloc %q: %v", r.task.Name, r.alloc.ID, err)
+			token = ""
+			goto OUTER
+		}
+
+		// The Vault token is valid now, so set it
+		r.vaultFuture.Set(token)
+
+		if updatedToken {
+			switch r.task.Vault.ChangeMode {
+			case structs.VaultChangeModeSignal:
+				s, err := signals.Parse(r.task.Vault.ChangeSignal)
+				if err != nil {
+					e := fmt.Errorf("failed to parse signal: %v", err)
+					r.logger.Printf("[ERR] client: %v", err)
+					r.Kill("vault", e.Error())
+					return
+				}
+
+				if err := r.Signal("vault", "new Vault token acquired", s); err != nil {
+					r.logger.Printf("[ERR] client: failed to send signal to task %v for alloc %q: %v", r.task.Name, r.alloc.ID, err)
+					r.Kill("vault", fmt.Sprintf("failed to send signal to task: %v", err))
+					return
+				}
+			case structs.VaultChangeModeRestart:
+				r.Restart("vault", "new Vault token acquired")
+			case structs.VaultChangeModeNoop:
+				fallthrough
+			default:
+				r.logger.Printf("[ERR] client: Invalid Vault change mode: %q", r.task.Vault.ChangeMode)
+			}
+
+			// We have handled it
+			updatedToken = false
+
+			// Call the handler
+			r.updatedTokenHandler()
+		}
+
+		// Start watching for renewal errors
+		select {
+		case err := <-renewCh:
+			// Clear the token
+			token = ""
+			r.logger.Printf("[ERR] client: failed to renew Vault token for task %v on alloc %q: %v", r.task.Name, r.alloc.ID, err)
+
+			// Check if we have to do anything
+			if r.task.Vault.ChangeMode != structs.VaultChangeModeNoop {
+				updatedToken = true
+			}
+		case <-r.waitCh:
+			return
+		}
+	}
+}
+
+// deriveVaultToken derives the Vault token using exponential backoffs. It
+// returns the Vault token and whether the token is valid. If it is not valid we
+// are shutting down
+func (r *TaskRunner) deriveVaultToken() (string, bool) {
+	attempts := 0
+	for {
+		tokens, err := r.vaultClient.DeriveToken(r.alloc, []string{r.task.Name})
+		if err != nil {
+			backoff := (1 << (2 * uint64(attempts))) * vaultBackoffBaseline
+			if backoff > vaultBackoffLimit {
+				backoff = vaultBackoffLimit
+			}
+			r.logger.Printf("[ERR] client: failed to derive Vault token for task %v on alloc %q: %v; retrying in %v", r.task.Name, r.alloc.ID, err, backoff)
+
+			attempts++
+
+			// Wait till retrying
+			select {
+			case <-r.waitCh:
+				return "", false
+			case <-time.After(backoff):
+			}
+		} else {
+			return tokens[r.task.Name], true
```

Comment:
[nitpick] I like putting simple early exit branches first so you can skip having an `else` clause altogether:

```go
tokens, err := ...
if err == nil {
  return tokens[r.task.Name], true
}

// implicit else
```

ymmv no big deal